### ISSUE TITLE
Ruby gem installer can consume all resources

### ISSUE THREAD

**Issue by Op3n4M3** [grpc/grpc#9861]
When installing the grpc gem the native extension compilation can consume all resources.

The installer is configured to use "make -j" this no count parameter which allows for jobs to fork for an unrestricted limit.  See link below.  When the gem in installed on a limited resource linux machine, for instance a raspberry pi, this can result in exhaustion of all memory available.

https://www.gnu.org/software/make/manual/html_node/Parallel.html#Parallel

**Comment by apolcyn** [comment]
Ideally when the grpc gem is installed the binary should come with it pre-compiled - I'm not sure exactly what's preventing that in this case. What's the `RUBY_PLATFORM` here? Also do both `gem install grpc` and `bundle install grpc` try to compile from source?

**Comment by Op3n4M3** [comment]
I have performed both, the target is a raspberry pi on arm architecture and does not include a pre-compiled binary.  When make is called by the gem installer for native compilation it exhausts all resources.  As offered in the PR setting reasonable job counts can make the gem more portable.

Note, I have been able to manually execute the compile inside the gem cache with some configuration work, however due to the method used to execute the make in this package I have been unable control the job count to get the gem to fully installed.  I am looking to use the grpc gem as a dependency for another gem I would like to be capable of installing on embedded platforms.

**Comment by apolcyn** [comment]
Overall I'm good with the fix but it would be nice to check that this isn't a problem with the grpc gem itself. Mainly interested in the `RUBY_VERSION` and `RUBY_PLATFORM` that this is happening on. 
FTR there was a previous issue where `gem install grpc` on OS X incorrectly tried to compile from source on install, which was fixed with an upgraded `rubygems` (https://github.com/grpc/grpc/issues/7727)

**Comment by Op3n4M3** [comment]
Understood, here are the details from the install I am trying to complete.  Can provide additional info if need.  As far as I can tell arm architecture is not part of the current gem cross compile list.

See Rakefile
```
  ext.cross_platform = [
    'x86-mingw32', 'x64-mingw32',
    'x86_64-linux', 'x86-linux',
    'universal-darwin'
  ]
```

Environment details:

```
$ uname -a
Linux ubuntu 3.18.0-20-rpi2 #21-Ubuntu SMP PREEMPT Sun Apr 5 01:56:02 UTC 2015 armv7l armv7l armv7l GNU/Linux

$ irb
2.3.3 :001 > RUBY_VERSION
 => "2.3.3" 
2.3.3 :002 > RUBY_PLATFORM
 => "armv7l-linux-eabihf" 

$ gem --version
2.5.2
```
Also attempted to update rubygems and called the installs again:
```
$ gem update --system
Updating rubygems-update
Successfully installed rubygems-update-2.6.10
Parsing documentation for rubygems-update-2.6.10
Installing darkfish documentation for rubygems-update-2.6.10
Done installing documentation for rubygems-update after 586 seconds
Parsing documentation for rubygems-update-2.6.10
Done installing documentation for rubygems-update after 0 seconds
Installing RubyGems 2.6.10
RubyGems 2.6.10 installed
Parsing documentation for rubygems-2.6.10
...

$ gem --version
2.6.10
ubuntu@ubuntu:~/metasploit-aggregator/ruby$ gem install grpc
Building native extensions.  This could take a while...

$ bundle install
Fetching gem metadata from https://rubygems.org/........
Fetching version metadata from https://rubygems.org/.
Resolving dependencies.....
Using rake 10.5.0
Using public_suffix 2.0.5
Using bundler 1.14.2
Using coderay 1.1.1
Using diff-lcs 1.3
Using multipart-post 2.0.0
Using google-protobuf 3.2.0
Using jwt 1.5.6
Using little-plugger 1.1.4
Using multi_json 1.12.1
Using memoist 0.15.0
Using os 0.9.6
Using grpc-tools 1.1.2
Using method_source 0.8.2
Using slop 3.6.0
Using rspec-support 3.5.0
Using addressable 2.5.0
Using faraday 0.11.0
Using logging 2.1.0
Using pry 0.10.4
Using rspec-core 3.5.4
Using rspec-expectations 3.5.0
Using rspec-mocks 3.5.0
Using signet 0.7.3
Using rspec 3.5.0
Using googleauth 0.5.1
Installing grpc 1.1.2 with native extensions
Killed
$
```

Results from ```top``` while bundler is running (free memory is quickly consumed):

```
top - 06:33:52 up 2 days, 12:45,  2 users,  load average: 38.85, 46.05, 21.29
Tasks: 299 total,  93 running, 205 sleeping,   0 stopped,   1 zombie
%Cpu(s): 87.8 us, 12.2 sy,  0.0 ni,  0.0 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem:    946112 total,   530088 used,   416024 free,    11204 buffers
KiB Swap:        0 total,        0 used,        0 free.    63136 cached Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                            
29446 ubuntu    20   0   23488  17152   9600 R  35.2  1.8   0:01.78 cc1                                
29415 ubuntu    20   0       0      0      0 R  24.5  0.0   0:01.30 cc1                                
29418 ubuntu    20   0   20148  14972   9780 R  24.5  1.6   0:01.28 cc1                                
29438 ubuntu    20   0   20516  15400   9548 R  23.2  1.6   0:01.25 cc1                                
29412 ubuntu    20   0   22400  15780   9604 R  22.2  1.7   0:01.29 cc1                                
29429 ubuntu    20   0   20140  14456   9416 R  21.9  1.5   0:01.20 cc1                                
29455 ubuntu    20   0   22336  15692   9748 R  20.9  1.7   0:01.13 cc1                                
28928 ubuntu    20   0    7572   5332   1284 D   3.6  0.6   0:06.16 make                               
29694 ubuntu    20   0   17464   9148   6908 R   2.3  1.0   0:00.07 cc1                                
29765 ubuntu    20   0   17332   8088   6332 R   2.0  0.9   0:00.06 cc1                                
29780 ubuntu    20   0   17332   8136   6376 R   2.0  0.9   0:00.06 cc1                                
29793 ubuntu    20   0   17332   8060   6308 R   2.0  0.9   0:00.06 cc1                                
29890 ubuntu    20   0   17332   8360   6316 R   2.0  0.9   0:00.06 cc1                                
29892 ubuntu    20   0   17332   8240   6460 R   2.0  0.9   0:00.06 cc1                                
29907 ubuntu    20   0   17332   8336   6292 R   2.0  0.9   0:00.06 cc1                                
29911 ubuntu    20   0   17332   8384   6340 R   2.0  0.9   0:00.06 cc1                                
29914 ubuntu    20   0   17332   8132   6372 R   2.0  0.9   0:00.06 cc1                                
29932 ubuntu    20   0   17332   8396   6384 R   2.0  0.9   0:00.06 cc1                                
26211 ubuntu    20   0    4832   1352    780 R   1.6  0.1   0:06.30 top                                
29575 ubuntu    20   0   17332   7824   6320 R   1.6  0.8   0:00.05 cc1                                
29602 ubuntu    20   0   17332   8180   6400 R   1.6  0.9   0:00.06 cc1                                
29617 ubuntu    20   0   17332   8028   6280 R   1.6  0.8   0:00.05 cc1                                
29629 ubuntu    20   0   17332   8144   6368 R   1.6  0.9   0:00.05 cc1                                
29654 ubuntu    20   0   17464   8144   6380 R   1.6  0.9   0:00.05 cc1                                
29660 ubuntu    20   0   17332   8072   6308 R   1.6  0.9   0:00.05 cc1                                
29663 ubuntu    20   0   17332   8080   6312 R   1.6  0.9   0:00.05 cc1                                
29671 ubuntu    20   0   17332   8284   6504 R   1.6  0.9   0:00.05 cc1                                
29681 ubuntu    20   0   17332   8112   6332 R   1.6  0.9   0:00.05 cc1                                
29685 ubuntu    20   0   17332   8196   6424 R   1.6  0.9   0:00.05 cc1                                
29707 ubuntu    20   0   17332   7980   6232 R   1.6  0.8   0:00.05 cc1                                
29710 ubuntu    20   0   17332   8116   6364 R   1.6  0.9   0:00.05 cc1         
```

**Comment by Op3n4M3** [comment]
Resource limitation released in v1.2.0.
